controller.py

```diff
@@ -138,3 +148,50 @@ def search(self, ip_address=None):
                 3600 * 24 * 30
             )
             return Response(body, 200, headers)
+
+    def register(self, do_get=HTTP.get_with_timeout):
+        opds_url = flask.request.form.get("url")
+        if not opds_url:
+            return NO_OPDS_URL
+
+        try:
+            response = do_get(opds_url, allowed_response_codes=["2xx", "3xx"])
+            feed = feedparser.parse(response.content)
+        except Exception, e:
+            return INVALID_OPDS_FEED
+        links = feed.get("feed", {}).get("links", [])
+
+        auth_url = None
+        for link in links:
+            # TODO: Define a rel for the auth document itself, and
+            # look for that instead.
```

Comment:
The link relation exists, it's `http://opds-spec.org/auth/document`. I suspect Hadrien added it to the A4OPDS spec recently, because I remember planning to define it and being surprised that it was already in there.

Forcing a 401 by getting the shelf is a good fallback if there is no http://opds-spec.org/auth/document. So here's what I'd do:

1. GET the registration URL.
2. If it's a 401, read the auth document from the body
3. Otherwise, GET the auth document.
4. If no link to auth document, GET the shelf link and hope for a 401.
5. Give up and complain.